memory.c

```diff
@@ -0,0 +1,329 @@
+    if (pfn < 0 || pfn >= NUM_VPN) { // Assuming NUM_VPN is the total number of physical frames
+        TracePrintf(0, "free_frame: ERROR: Invalid physical frame number %d\n", pfn);
+        return;
+    }
+
+    // Check if the frame was actually allocated before freeing
+    if (frame_bitMap[pfn] == 0) {
+        TracePrintf(0, "free_frame: WARNING: Attempted to free an already free frame %d\n", pfn);
+        return;
+    }
+
+    frame_bitMap[pfn] = 0; // Mark the frame as free
+    TracePrintf(5, "free_frame: Freed frame %d\n", pfn);
+}
+
+int SetKernelBrk(void *addr) {
+    if (!vm_enabled) {
+        // Before VM enabled: just track the kernel break
+        if (addr > kernel_brk && addr < (void *)KERNEL_STACK_BASE) {
+            kernel_brk = addr;
+            return 0;
+        } else {
+            return ERROR;
+        }
+    } else {
+        // After VM enabled: need to actually allocate frames and update page tables
+        void *old_brk = kernel_brk;
+        void *new_brk = addr;
+
+        // Round up to page boundary
+        new_brk = (void *)UP_TO_PAGE((unsigned int)new_brk);
+
+        // If decreasing break, just update the break pointer
+        if (new_brk <= old_brk) {
+            kernel_brk = new_brk;
+            return 0;
+        }
+
+        // If increasing break, map new frames
+        for (void *p = old_brk; p < new_brk; p += PAGESIZE) {
+            int vpn = (int)p >> PAGESHIFT;
+            int pfn = allocate_frame(); // Get a free physical frame
+
+            if (pfn == ERROR) {
+                return ERROR; // Out of physical memory
+            }
+
+            // Update the page table entry using explicit pointer arithmetic
+            pte_t *current_pte = region0_pt + vpn; // Pointer to the specific PTE
+            current_pte->valid = 1;
+            current_pte->prot = PROT_READ | PROT_WRITE;
+            current_pte->pfn = pfn;
+
+            // Flush the TLB for this page
+            WriteRegister(REG_TLB_FLUSH, (unsigned int)p);
+        }
+
+        kernel_brk = new_brk;
+        return 0;
+    }
+}
+
+
+
+void init_page_table(int kernel_text_start, int kernel_data_start, int kernel_brk_start) {
+    TracePrintf(0, "Initializing page table...\n");
+    // Set the base address for the region0 page table in hardware
+    unsigned int last_physical_address = MAX_PMEM_SIZE - 1; // Calculate the last physical address
+    unsigned int page_table_size = MAX_PT_LEN * sizeof(struct pte); // Calculate size needed for the page table
+    unsigned int p_address_region0_base = DOWN_TO_PAGE(last_physical_address - page_table_size); // Place page table at the end of physical memory
+    region0_pt = (pte_t *)p_address_region0_base; // Cast to pte_t*
+
+    // Initialize the kernel text, data, and heap sections in the page table
+    for (int vpn = kernel_text_start; vpn < kernel_brk_start; vpn++) {  //
+        pte_t *entry = (pte_t *)(p_address_region0_base + vpn * sizeof(pte_t));

```

Comment:
you can just use `pte_t *entry = &region0_pt[vpn]`

Reply:
this works because region0_pt is typed as a pte_t pointer, so therefore indexing into this will move it by the appropiate amount (sizeof pte_t)

Reply:
something in this loop is causing your segmentation fault, b/c you don't get to the next trace statement. hopefully fixing the location of the page table and the method for indexing into it can resolve this